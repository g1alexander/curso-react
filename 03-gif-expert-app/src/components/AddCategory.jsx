import { useState } from "react";
import PropTypes from "prop-types";

export function AddCategory({ setCategories }) {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (inputValue.trim().length > 2) {
      setCategories((categories) => [inputValue, ...categories]);
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input
        value={inputValue}
        onChange={handleInputChange}
        type="text"
        name="add"
        id="add"
      />
    </form>
  );
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
