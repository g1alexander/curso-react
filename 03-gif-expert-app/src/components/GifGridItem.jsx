import PropTypes from "prop-types";

export function GifGridItem({ title, image }) {
  return (
    <div className="card animate__animated animate__bounce">
      <h3>{title}</h3>
      <img src={image} alt={title} />
    </div>
  );
}

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
};
