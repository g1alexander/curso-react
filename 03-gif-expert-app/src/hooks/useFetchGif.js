import { useEffect, useState } from "react";
import { getGifApi } from "../helpers/getGifApi";

export function useFetchGif(category) {
  const [state, setState] = useState({
    data: [],
    loading: true,
  });

  useEffect(() => {
    getGifApi(category).then((data) => {
      setState({
        data,
        loading: false,
      });
    });
  }, [category]);

  return state;
}
