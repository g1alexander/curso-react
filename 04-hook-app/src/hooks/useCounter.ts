import { useState } from "react";

export function useCounter(initialValue = 10) {
  const [counter, setCounter] = useState<number>(initialValue);

  const increment = (payload: number): void => {
    setCounter((current) => current + payload);
  };

  const decrement = (payload: number): void => {
    setCounter((current) => current - payload);
  };

  const reset = (): void => {
    setCounter(initialValue);
  };

  return {
    counter,
    increment,
    decrement,
    reset,
  };
}
