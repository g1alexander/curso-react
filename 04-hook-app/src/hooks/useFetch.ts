import { useEffect, useState } from "react";

interface State<T> {
  data: T | null;
  isLoading: boolean;
  hasError: null | string;
}

export function useFetch<T = unknown>(url: string): State<T> {
  const [state, setstate] = useState<State<T>>({
    data: null,
    isLoading: true,
    hasError: null,
  });

  const getFetch = async () => {
    setstate({
      ...state,
      isLoading: true,
    });

    const response = await fetch(url);
    const data = await response.json();

    setstate({
      hasError: null,
      data,
      isLoading: false,
    });
  };

  useEffect(() => {
    getFetch();
  }, [url]);

  return {
    ...state,
  };
}
