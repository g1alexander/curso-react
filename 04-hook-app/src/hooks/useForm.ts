import { useState } from "react";

export function useForm<T>(initialState: T) {
  const [formState, setFormState] = useState<T>(initialState);

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;

    setFormState({
      ...formState,
      [name]: value,
    });
  };

  const handleResetForm = () => {
    setFormState(initialState);
  };

  return {
    ...formState,
    formState,
    handleInputChange,
    handleResetForm,
  };
}
