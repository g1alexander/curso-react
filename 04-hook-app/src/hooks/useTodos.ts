import { useEffect, useReducer } from "react";

import { type State, todoReducer, Action } from "../08-useReducer";

const initialState: State[] = [];

const init = () => {
  return JSON.parse(localStorage.getItem("todos") || "[]");
};

export const useTodos = () => {
  const [todos, dispatchTodo] = useReducer(todoReducer, initialState, init);

  useEffect(() => {
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  const onNewTodo = (todo: State) => {
    const action: Action = {
      type: "[TODO] add todo",
      payload: todo,
    };

    dispatchTodo(action);
  };

  const onDeleteTodo = (todo: State) => {
    const action: Action = {
      type: "[TODO] delete todo",
      payload: todo,
    };

    dispatchTodo(action);
  };

  const onToggleTodo = (todo: State) => {
    const action: Action = {
      type: "[TODO] toggle todo",
      payload: todo,
    };

    dispatchTodo(action);
  };

  const onPendingTodos = () => {
    return todos.filter((todo) => !todo.done).length;
  };

  const onCompletedTodos = () => {
    return todos.filter((todo) => todo.done).length;
  };

  return {
    todos,
    onNewTodo,
    onDeleteTodo,
    onToggleTodo,
    onPendingTodos,
    onCompletedTodos,
  };
};
