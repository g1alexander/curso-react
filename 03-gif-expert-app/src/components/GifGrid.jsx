import PropTypes from "prop-types";
import { GifGridItem } from "./GifGridItem";
import { useFetchGif } from "../hooks/useFetchGif";

export function GifGrid({ category }) {
  const { data: images, loading } = useFetchGif(category);

  return (
    <>
      <h2>{category}</h2>

      {loading && (
        <p className="animate__animated animate__flash">cargando...</p>
      )}

      <section className="grid">
        {images.map((img) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </section>
    </>
  );
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
