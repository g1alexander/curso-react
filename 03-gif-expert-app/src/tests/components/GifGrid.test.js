import { shallow } from "enzyme";
import { GifGrid } from "../../components/GifGrid";
import { useFetchGif } from "../../hooks/useFetchGif";
jest.mock("../../hooks/useFetchGif");

describe("pruebas del componente <GifGrid />", () => {
  const text = "the beatles";

  test("el componente se debe mostrar correctamente", () => {
    useFetchGif.mockReturnValue({
      data: [],
      loading: true,
    });
    const wrapper = shallow(<GifGrid category={text} />);

    expect(wrapper).toMatchSnapshot();
  });

  test("debe mostrarse items cuando se cargan imagenes useFetchGif", () => {
    const gifs = [
      {
        id: "ABC",
        title: "hola",
        image: "https://localhost/hola.png",
      },
    ];

    useFetchGif.mockReturnValue({
      data: gifs,
      loading: false,
    });

    const wrapper = shallow(<GifGrid category={text} />);

    expect(wrapper).toMatchSnapshot();

    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
  });
});
