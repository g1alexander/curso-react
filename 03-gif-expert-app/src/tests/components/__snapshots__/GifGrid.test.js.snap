// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas del componente <GifGrid /> debe mostrarse items cuando se cargan imagenes useFetchGif 1`] = `
<Fragment>
  <h2>
    the beatles
  </h2>
  <section
    className="grid"
  >
    <GifGridItem
      id="ABC"
      image="https://localhost/hola.png"
      key="ABC"
      title="hola"
    />
  </section>
</Fragment>
`;

exports[`pruebas del componente <GifGrid /> el componente se debe mostrar correctamente 1`] = `
<Fragment>
  <h2>
    the beatles
  </h2>
  <p
    className="animate__animated animate__flash"
  >
    cargando...
  </p>
  <section
    className="grid"
  />
</Fragment>
`;
